Use clearInterval to wait for React before init

paulmolluzzo opened the pull request:
This is a super simple solution to waiting for React before firing the `init` function. Since the `init` only really depends on the `header` element, we're waiting on that to load into the DOM.

This fixes #15.

// @sindresorhus

paulmolluzzo committed:
Use clearInterval to wait for React updates before firing functions

sindresorhus commented on browser.js:
```diff
@@ -310,6 +310,11 @@ document.addEventListener('DOMContentLoaded', () => {
+			init();
+			clearInterval(awaitReact);
+		}
+	},100)
```
```
}, 100);
```

sindresorhus commented on browser.js:
```diff
@@ -310,6 +310,11 @@ document.addEventListener('DOMContentLoaded', () => {
-	// TODO: figure out a better way to detect when React is done
-	setTimeout(init, 300);
+	// detect when React is ready before firing init
+	const awaitReact = setInterval(() => {
```
Maybe make this reusable by making it a `waitsFor` function that accepts a selector? We need this for other things too, see the other TODO's.

paulmolluzzo replied:
Good call. Do you want to have it support arrays/NodeLists or just individual nodes?

sindresorhus replied:
Both I guess, can just do `$(sel).get(0)`

paulmolluzzo replied:
This is using `querySelector` so it's only getting the first matching element, and we (currently) have no reason to ask for more. I would only say that we're looking for _at least one_ of some item ever anyway. I can't think of any scenario where you'd need a whole list of nodes. For example, you won't click more than 1 thing at a time. I could change it to a `querySelectorAll`, but I don't see why we need that. Feel free to point out something I'm missing. 😄

paulmolluzzo commented:
@sindresorhus I've updated this as suggested, and updated/replaced the `newTweet` and `newDM` functions as well.

---

One small note that's sort of tangential. In this block, you only need to click for messages if you're not on the messages page:

``` js
function newDM() {
    $('a[href$="/messages"]').click();
.
.
.
}`
```

Shouldn't this `$('a[href$="/messages"]').click();` only be done on non-messages pages?  (i.e. something like `/^\/(^messages)/.test(pathname)`)

I can fix in this PR, but it's not _technically_ part of the issue we're solving, if that matters.

paulmolluzzo commented:
Made the branch up to date with `master`.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,35 @@ function changeTab(next) {
 
-	$('a[href$="/compose/tweet"]').click();
+	// wait for new tweet button to click it
+	waitFor('a[href$="/compose/tweet"]', () => {
```
Let's make `waitFor` return a Promise instead of a callback. And let the Promise return the DOM element. Then it could just be:

``` js
waitFor('a[href$="/compose/tweet"]').then(el => el.click());
```

paulmolluzzo replied:
Good idea.

sindresorhus commented:
> Shouldn't this $('a[href$="/messages"]').click(); only be done on non-messages pages?

Yes, good point. Just do that in a follow-up commit after this, as it's a bit out of scope of this PR.

paulmolluzzo committed:
Promisify waitFor function

paulmolluzzo commented:
Pushed a promisify update but didn't rebase (just squash later if you want). This seems slower to me from a UX perspective though. Could be my imagination.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,32 @@ function changeTab(next) {
+function waitFor(selectorString) {
+	return new Promise(resolve => {
+		const awaitReact = setInterval(() => {
+			if ($(selectorString)) {
```
You're getting the element twice here. Instead get the element, assign to variable, check variable, and then pass that variable to resolve.

paulmolluzzo replied:
Might have been the reason for the slight lag.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,32 @@ function changeTab(next) {
+// sets interval to wait for selector to be ready before firing callback
+function waitFor(selectorString) {
+	return new Promise(resolve => {
+		const awaitReact = setInterval(() => {
```
Shouldn't be React specific, so I would rather use `awaitElement`.

paulmolluzzo replied:
Changed.

sindresorhus commented:
> This seems slower to me from a UX perspective though.

Weird, I'll do some tests tomorrow myself.

paulmolluzzo commented:
> Weird, I'll do some tests tomorrow myself.

I'll take a closer look too, but it might have been the double selecting. Technically we used to call the function inside the interval, now we resolve the Promise and then call the function, so we've at least added the promise resolution process in there. I don't know if creating the promise itself made a big impact.

If you run performance tests of any kind please update regardless. I'm interested in how to best test performance.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,33 @@ function changeTab(next) {
+// sets interval to wait for selector to be ready before firing callback
+function waitFor(selectorString) {
+	return new Promise(resolve => {
+		const awaitElement = setInterval(() => {
```
Let's also do the check before the interval in case it already exists before the check started, so we don't have to wait the 100ms.

paulmolluzzo replied:
To do this outside of the `setInterval` I have to turn this into a function. A regular assignment won't change on each interval, causing an infinite loop of the initial state on load.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,33 @@ function changeTab(next) {
+				resolve(selectedElement);
+				clearInterval(awaitElement);
+			}
+		}, 100);
```
Let's make this 50.

paulmolluzzo commented:
@sindresorhus I rebased on the commit to promisify this thing with the changes you suggested. This is the fastest of the 3 variations, clearly because of the "shortcut" and the reduced interval time. You hit that shortcut nicely on new tweets, and when we fix that other thing for messages (the URL regex) that will get a little faster too.

Thanks for the feedback. Any other improvements we can make are welcome.

sindresorhus commented on browser.js:
```diff
@@ -98,27 +98,40 @@ function changeTab(next) {
+			}
+		}, 50);
+	});
+}
```
I would go with:

``` js
// sets interval to wait for selector to be ready before firing callback
function waitFor(selector) {
    return new Promise(resolve => {
        const el = $(selector);

        // shortcut if the element already exists
        if (el) {
            resolve(el);
            return;
        }

        // interval to keep checking for it to come into the DOM
        const interval = setInterval(() => {
            const el = $(selector);

            if (el) {
                resolve(el);
                clearInterval(interval);
            }
        }, 50);
    });
}
```

paulmolluzzo replied:
I think you just like short variable names. 😄 

All pushed.
